docs/design/enums_or_ids.md

```diff
@@ -0,0 +1,92 @@
+Enums or Identifiers
```

Comment:
I understand the topic that this is addressing, but the naming and definitions could help.

I think the topic is about representing values -- specifically, when the range of values is finite, and the values form a discrete set.  So, a finite set of discrete values.

I'd say that an entity is a standalone thing. Instead of the word "entity", I would use the word "element" instead to represent an item in a set -- both for the CS and the math definitions of "set".

I agree on the definition of identifier as a space of strings consisting of numbers, letters, etc. that stand in for something. (Some languages define "keyword" as a language-reserved type of identifier with specific semantics, and leave "identifier" for user-defined constructs. Other languages use "symbol" to represent the entire space of such syntactic token strings of keywords+identifiers.)  

But long story short, I don't think "identifier" is the right word to use (as a subcategory of this larger topic, for which enum is one alternative).

Integers and strings are used as examples for what is called "identifier". But I don't think we would use a floating point or a fixed decimal in such examples. So I think the key characteristic is that integers and strings are _unbounded_ discrete values. 

The unbounded nature of integers and strings gets to the heart of the problem that the decision tree questions address -- (are we dealing with discrete values? is the range a finite set?)  what is the size (cardinality) of the set of values? can the set grow? 

"Unbounded discrete value type" isn't concise like "identifier", but they do feel very distinct to me. You could also just say "ints/strings", that sounds fine to me.  The newtype idiom seems like a wrapper around some primitive type, from what I've seen of it so far.

Reply:
Thanks for the feedback.  Responses:

> I understand the topic that this is addressing, but the naming and definitions could help.
> 
> I think the topic is about representing values -- specifically, when the range of values is finite, and the values form a discrete set. So, a finite set of discrete values.

The discrete set is not necessarily finite. In many cases, it may be *de facto* finite but theoretically infinite.

> I'd say that an entity is a standalone thing. Instead of the word "entity", I would use the word "element" instead to represent an item in a set -- both for the CS and the math definitions of "set".

Part of my reasoning in choosing the word "entity" was because I wanted to emphasize that the thing in question has standalone value.  For example, it is meaningful to talk about a certain language or a certain time zone outside the context of talking about the other elements of the set.  I will clarify that in the doc.

> I agree on the definition of identifier as a space of strings consisting of numbers, letters, etc. that stand in for something. (Some languages define "keyword" as a language-reserved type of identifier with specific semantics, and leave "identifier" for user-defined constructs. Other languages use "symbol" to represent the entire space of such syntactic token strings of keywords+identifiers.)
> 
> But long story short, I don't think "identifier" is the right word to use (as a subcategory of this larger topic, for which enum is one alternative).

Part of my inspiration for choosing the terminology I chose is UTS 39:

http://unicode.org/reports/tr39/#Identifier_Characters

From that spec: "Identifiers ("IDs") are strings used in application contexts to refer to specific entities of certain significance in the given application. In a given application, an identifier will map to at most one specific entity."

I will add a reference to the spec in the doc.

> Integers and strings are used as examples for what is called "identifier". But I don't think we would use a floating point or a fixed decimal in such examples. So I think the key characteristic is that integers and strings are _unbounded_ discrete values.
>
> The unbounded nature of integers and strings gets to the heart of the problem that the decision tree questions address -- (are we dealing with discrete values? is the range a finite set?) what is the size (cardinality) of the set of values? can the set grow?
> 
> "Unbounded discrete value type" isn't concise like "identifier", but they do feel very distinct to me. You could also just say "ints/strings", that sounds fine to me. The newtype idiom seems like a wrapper around some primitive type, from what I've seen of it so far.

An identifier is not necessarily unbounded.  There are only 256 possible entries if a u8 is used, for example.  The identifier can also specify its own syntax of valid values.  For example, there are only 26^2 possible ISO 639-1 language subtags.

The reason we would most likely not use a floating point number in this context is because floating point numbers lack precision, and there may be multiple bit patterns for the same value (e.g., NaN has many bit patterns).  I imagine that a fixed decimal could be used as an identifier, but more likely you would use a string containing a fixed decimal, since strings are more widely accepted as a medium of communication between programs.

Reply:
Ah, okay, I think I understand better now, the space of string values that you are considering are those that would also qualify as UTS 39 identifiers. And you are using the word "identifier" in that sense. That seems reasonable to me.

And I see the point about how the space of identifiers are not necessarily infinite, but for our purposes (memory/data size considerations), often times it might be.

The main point I was making about finite/bounded vs. unbounded is that enums are finite and discrete, so to have the identifiers be an alternative for situations that the decision tree describes, then for whatever identity / value under consideration, the set of valid values is always something that is finite and discrete. That set of values may expand, and that's the point of the doc.

But I think my previous comment went off on a tangent b/c I didn't understand what "identifier" was representing.